pkg/operator/controller/controller_dns_daemonset.go

```diff
@@ -0,0 +1,186 @@
+	for i, c := range daemonset.Spec.Template.Spec.Containers {
+		switch c.Name {
+		case "dns":
+			daemonset.Spec.Template.Spec.Containers[i].Image = r.CoreDNSImage
+		case "dns-node-resolver":
+			daemonset.Spec.Template.Spec.Containers[i].Image = r.OpenshiftCLIImage
+			envs := []corev1.EnvVar{}
+			if len(clusterIP) > 0 {
+				envs = append(envs, corev1.EnvVar{
+					Name:  "NAMESERVER",
+					Value: clusterIP,
+				})
+			}
+			if len(clusterDomain) > 0 {
+				envs = append(envs, corev1.EnvVar{
+					Name:  "CLUSTER_DOMAIN",
+					Value: clusterDomain,
+				})
+			}
+
+			if daemonset.Spec.Template.Spec.Containers[i].Env == nil {
+				daemonset.Spec.Template.Spec.Containers[i].Env = []corev1.EnvVar{}
+			}
+			daemonset.Spec.Template.Spec.Containers[i].Env = append(daemonset.Spec.Template.Spec.Containers[i].Env, envs...)
+		}
+	}
+	return daemonset, nil
+}
+
+// currentDNSDaemonSet returns the current dns daemonset.
+func (r *reconciler) currentDNSDaemonSet(dns *operatorv1.DNS) (*appsv1.DaemonSet, error) {
+	daemonset := &appsv1.DaemonSet{}
+	if err := r.client.Get(context.TODO(), DNSDaemonSetName(dns), daemonset); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return daemonset, nil
+}
+
+// createDNSDaemonSet creates a dns daemonset.
+func (r *reconciler) createDNSDaemonSet(daemonset *appsv1.DaemonSet) error {
+	if err := r.client.Create(context.TODO(), daemonset); err != nil {
+		return fmt.Errorf("failed to create dns daemonset %s/%s: %v", daemonset.Namespace, daemonset.Name, err)
+	}
+	logrus.Infof("created dns daemonset: %s/%s", daemonset.Namespace, daemonset.Name)
+	return nil
+}
+
+// updateDNSDaemonSet updates a dns daemonset.
+func (r *reconciler) updateDNSDaemonSet(current, desired *appsv1.DaemonSet) error {
+	changed, updated := daemonsetConfigChanged(current, desired)
+	if !changed {
+		return nil
+	}
+
+	if err := r.client.Update(context.TODO(), updated); err != nil {
+		return fmt.Errorf("failed to update dns daemonset %s/%s: %v", updated.Namespace, updated.Name, err)
+	}
+	logrus.Infof("updated dns daemonset: %s/%s", updated.Namespace, updated.Name)
+	return nil
+}
+
+// daemonsetConfigChanged checks if current config matches the expected config
+// for the cluster dns daemonset and if not returns the updated config.
+func daemonsetConfigChanged(current, expected *appsv1.DaemonSet) (bool, *appsv1.DaemonSet) {
+	changed := false
+	updated := current.DeepCopy()
+	for _, name := range []string{"dns", "dns-node-resolver"} {
+		var curIndex int
+		var curImage, expImage string
+
+		for i, c := range current.Spec.Template.Spec.Containers {
+			if name == c.Name {
+				curIndex = i
+				curImage = current.Spec.Template.Spec.Containers[i].Image
+				break
+			}
+		}
+		for i, c := range expected.Spec.Template.Spec.Containers {
+			if name == c.Name {
+				expImage = expected.Spec.Template.Spec.Containers[i].Image
+				break
+			}
+		}
+
+		if len(curImage) == 0 {
+			return true, expected
```

Comment:
How can `curImage` be empty? Should we add a error return value for this contingency, or at least log it?

Reply:
We do not expect this to happen, added logging error in this case.

Reply:
The logging is an improvement.  However, I think that returning `expected` to `updateDNSDaemonSet` will cause the latter to fail when it tries to update the daemonset, because `expected` came from `desiredDNSDaemonSet` and does not have the resource version etc., right?  Probably the best thing to do here is to copy the full containers definition from `expected` to `updated` and then return `updated`:
```go
		if len(curImage) == 0 {
			logrus.Errorf("current daemonset %s/%s did not contain expected %s container", current.Namespace, current.Name, name)
			updated.Spec.Template.Spec.Containers = expected.Spec.Template.Spec.Containers
			changed = true
			break
```
Setting `changed = true`and using `break` means we continue with the usual control flow, which means we'll get the logic for updating volumes too, if and when we implement that, before we reach the final `return true, updated`.

Reply:
fixed